docs/design/proposal-troubleshoot-consolidated-cli.md

```diff
@@ -0,0 +1,350 @@
+
+### Public APIs
+
+The existing package entrypoint we use from the support-bundle command is `supportbundle.CollectSupportBundleFromSpec()`
+which is an end-to-end function that runs collectors, redactors and then analysis.
+
+The Consolidated cli should be responsible for handling the end-to-end running of this process in the `cmd` leaving the importable packages free to focus on defining the individual collect, redact analyze steps.
+
+To this end we should create a new `pkg` that targets the functionality provided by the `collect` and `analyze` packages such as `collect.runHostCollectors()` and provide a stable API to be used by the CLI and other projects. At the same time we can unpublish functionality that should not be exposed (such as the individual collector and analyser functions) and mark code for deprecation.
+
+This new package should be kept as minimal as possible. and serve only as an interface to private functions in the other packages.
+
+Once the stable API is ready we can instruct projects like kurl to target that and work on removing code marked for deprecation and migrate non-public functions to an `internal` package to make it clear that it's not intended to be imported.
+
+The core of this api will be exposed via new troubleshoot `types`
+
+```go
+type TroubleshootKinds struct {
+}
+
+type Bundle struct {
+  FilePath string
+}
+
+type AnalysisResults struct {
+  Bundle Bundle // include bundle metadata
+}
+
+```
+
+from there we can build out methods for each to interact with them:
+
+```go
+// Load loads specs defined by the LoadOptions struct and returns a TroubleshootKinds struct
+func Load(LoadOptions) (TroubleshootKinds,err){}
+
+// Bundle.Collect runs collections defined in TroubleshootKinds passed through CollectOptions
+func (bundle *Bundle) Collect(CollectOptions) (error) {}
+
+// Bundle.Load will create a Bundle struct from options in LoadBundleOptions
+func (bundle *Bundle) Load(LoadBundleOptions) (error) {}
+// it's worth noting that while this may appear inefficient now
+// it allows us to extend what we include in the Bundle struct in future without having to continuously extend the load function.
+
+// Bundle.Analyze runs analysis defined in TroubleshootKinds passed through AnalyzeOptions
+func (bundle *Bundle) Analyze(AnalyzeOptions) (AnalysisResults, error) {}
+
+// Bundle.Redact runs redaction defined in TroubleshootKinds passed through RedactOptions
+func (bundle *Bundle) Redact(RedactOptions) error {}
+
+// Bundle.Archive produces an archive from a bundle on disk with options passed in ArchiveOptions
+func (bundle *Bundle) Archive(ArchiveOptions) error {}
+
+// Bundle.Serve starts an sbctl-like server with options defined in ServeOptions
+func (bundle *Bundle) Serve(ServeOptions) error {}
+
+// AnalysisResults.Save saves the results of the analysis according to SaveAnalysisOptions
+func (results *AnalysisResults) Save(SaveAnalysisOptions) error {}
```

Comment:
Is this meant to save the results in an existing bundle?

Just wondering if `func (bundle *Bundle) Analyze(AnalyzeOptions) (AnalysisResults, error) {}` does this already, or whether it should save to a bundle.

Reply:
hmm I was thinking it would save the results to a file in a specified format, but I wonder if we should leave that to the implementing application instead?

Reply:
Yer, I see what you mean. There are a number of use cases here.

* Analyse and save results in the bundle
* Re-analyse an existing bundle but do not modify the bundle (e.g after loading an existing bundle which has results already, you may want to retain the old ones as-is)

Maybe these decisions can be driven by `AnalyzeOptions` e.g

```go
type AnalyzeOptions struct {
	Specs TroubleshootKinds // list of specs to extract analysers from
	SaveToBundle bool // save the results to a bundle. Or should we just check if BundlePath is empty and infer from that?
	BundlePath string // path to save the bundle to
}
```